core/src/replay_stage.rs

```diff
@@ -3046,6 +3047,51 @@ impl ReplayStage {
                     }
                 }
 
+                if bank.collector_id() != my_pubkey {
+                    // If the block does not have at least DATA_SHREDS_PER_FEC_BLOCK shreds in the last FEC set,
+                    // mark it dead. No reason to perform this check on our leader block.
+                    if !blockstore
+                        .is_last_fec_set_full(bank.slot())
+                        .inspect_err(|e| {
+                            warn!(
+                                "Unable to determine if last fec set is full for slot {} {},
+                                marking as dead: {e:?}",
+                                bank.slot(),
+                                bank.hash()
+                            )
+                        })
+                        .unwrap_or(false)
+                    {
+                        // Update metric regardless of feature flag
+                        datapoint_warn!(
+                            "incomplete_final_fec_set",
+                            ("slot", bank_slot, i64),
+                            ("hash", bank.hash().to_string(), String)
+                        );
+                        if bank
+                            .feature_set
+                            .is_active(&solana_sdk::feature_set::vote_only_full_fec_sets::id())
+                        {
+                            Self::mark_dead_slot(
+                                blockstore,
+                                bank,
+                                bank_forks.read().unwrap().root(),
```

Comment:
I think this holds the lock on back_forks for the entirety of this function call. Is that intentional?

Reply:
I don't believe it does since we copy the root, the `RwLockReadGuard` is dropped before we enter the function.
To validate I added an extra slot argument to the function and was able to get this to compile and pass tests:
```
Self::mark_dead_slot(
  blockstore,
  bank,
  bank_forks.read().unwrap().root(),
  bank_forks.write().unwrap().root(),
 ...
```

Reply:
are you sure this code is invoked in tests?
I am testing this here:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d148074d575f3c38c66156ce2ffd76c7
and it deadlocks. If you change the `write` to `read` it works which confirms it is a deadlock.

Reply:
you're right it seems this code wasn't being invoked in tests. 
took a look at `mark_dead_slot`, there is nothing here that requires the `bank_forks` lock. I've modified the code to not hold the lock.

I can do this for the other invocations of `mark_dead_slot` in a follow up pr